quint/src/flattening/flattener.ts

```diff
@@ -0,0 +1,163 @@
+ *
+ * @author Gabriela Moreira
+ *
+ * @module
+ */
+
+import { Definition, LookupTable, builtinNames } from '../names/base'
+import {
+  QuintApp,
+  QuintDef,
+  QuintExport,
+  QuintImport,
+  QuintInstance,
+  QuintModule,
+  QuintName,
+  qualifier,
+} from '../ir/quintIr'
+
+import { IRVisitor, walkDefinition, walkModule } from '../ir/IRVisitor'
+import { addNamespaceToDefinition } from '../ir/namespacer'
+
+/**
+ * Flatten a module, replacing instances, imports and exports with definitions refered by the module.
+ *
+ * @param quintModule - The module to be flattened
+ * @param modulesByName - A map of refered modules by name
+ * @param lookupTable - The lookup table for the module and all its references
+ *
+ * @returns The flattened module
+ */
+export function flattenModule(
+  quintModule: QuintModule,
+  modulesByName: Map<string, QuintModule>,
+  lookupTable: LookupTable
+): QuintModule {
+  const moduleCopy: QuintModule = { ...quintModule, declarations: [...quintModule.declarations] }
+  const flattener = new Flatenner(modulesByName, lookupTable)
+  walkModule(flattener, moduleCopy)
+  return moduleCopy
+}
+
+class Flatenner implements IRVisitor {
+  private modulesByName: Map<string, QuintModule>
+  private lookupTable: LookupTable
+  // Buffer of definitions to add to the module. We can try to make this ordered in the future.
+  // For now, we rely on toposorting defs after flattening.
+  private defsToAdd: Map<string, QuintDef> = new Map()
+  // Store a namespace to use when recursing into nested definitions
+  private namespaceForNested?: string
+
+  constructor(modulesByName: Map<string, QuintModule>, lookupTable: LookupTable) {
+    this.modulesByName = modulesByName
+    this.lookupTable = lookupTable
+  }
+
+  enterModule(_quintModule: QuintModule) {
+    // Reset `defsToAdd`
+    this.defsToAdd = new Map()
+  }
+
+  exitModule(quintModule: QuintModule) {
+    // Get rid of imports and exports, and add the definitions we collected
+    quintModule.declarations = quintModule.declarations.filter(d => d.kind !== 'import' && d.kind !== 'export')
+    quintModule.declarations.push(...this.defsToAdd.values())
+  }
+
+  enterName(name: QuintName) {
+    this.flattenName(name.id)
+  }
+
+  enterApp(app: QuintApp) {
+    this.flattenName(app.id)
+  }
+
+  enterExport(decl: QuintExport) {
+    // Find all top-level definitions in the exported module that are used somewhere
+    // (not necessarily in the module itself)
+    const ids = this.modulesByName.get(decl.protoName)!.declarations.map(d => d.id)
+    const definitions = [...this.lookupTable.values()].filter(d => ids.includes(d.id))
+
+    definitions.forEach(def => {
+      if (def.kind === 'param') {
+        throw new Error(
+          `Impossible: instersection of top-level declarations with lookup table entries should never be a param. Found ${def}`
+        )
+      }
+
+      if (this.defsToAdd.has(def.name)) {
+        // Already added
+        return
+      }
```

Comment:
This code looks a bit dangerous to me. We can accidentally ignore name collisions like that. Should we at least compare the signatures of the definitions, just to make sure that we are ignoring a duplicate definition that arrived via a different import path?

Reply:
That is already checked during name resolution. It should be impossible to have any name collisions here. I don't think it makes sense to write additional arbitrary checks when we already have a proper pass to check that - it would be the same as checking types after the type checker.

Reply:
Ah, makes sense to me